noxfile.py

```diff
@@ -18,16 +20,146 @@
+        hexdigest = hashlib.sha256(fi.read()).hexdigest()
+    tmp_dir = Path(session.create_tmp())
+    cache = tmp_dir / f"py{python_version}.sha"
+    with open(cache, "w") as fo:
+        fo.write(hexdigest)
+
+
+def combine_requirements(primary, secondary):
+    """
+    Combine the conda environment YAML files together into one.
+
+    Parameters
+    ----------
+    primary: str
+        The filename of the primary YAML conda environment.
+    secondary: str
+        The filename of the subordinate YAML conda environment.
+
+    Returns
+    -------
+    dict
+        A dictionary of the combined YAML conda environments.
+
+    """
+    with open(primary, "r") as fi:
+        result = yaml.load(fi, Loader=yaml.FullLoader)
+    with open(secondary, "r") as fi:
+        secondary = yaml.load(fi, Loader=yaml.FullLoader)
+    # Combine the channels and dependencies only.
+    for key in ["channels", "dependencies"]:
+        result[key] = sorted(set(result[key]).union(secondary[key]))
+    # Filter out iris as a dependency as this is dealt with separately.
+    result["dependencies"] = [
+        spec for spec in result["dependencies"] if not spec.startswith("iris")
+    ]
+    return result
+
+
+def get_iris_github_artifact(session):
+    """
+    Determine whether an Iris source artifact from GitHub is required.
+
+    This can be an Iris branch name, commit sha or tag name.
+
+    Parameters
+    ----------
+    session: object
+        A `nox.sessions.Session` object.
+
+    Returns
+    -------
+    str
+        The Iris GitHub artifact.
+
+    """
+    result = IRIS_SOURCE
+    # The CLI overrides the environment variable.
+    for arg in session.posargs:
+        if arg.startswith("--iris="):
+            parts = arg.split("=")
+            if len(parts) == 2:
+                result = parts[1].strip()
+                break
+    if result:
+        parts = result.strip().split(":")
+        result = None
+        if len(parts) == 2:
+            repo, artifact = parts
+            if repo.startswith("'") or repo.startswith('"'):
+                repo = repo[1:]
+            if repo.lower() == "github":
+                result = artifact
+                if result.endswith("'") or result.endswith('"'):
+                    result = result[:-1]
+    return result
+
+
+def venv_cached(session, fname):
+    """
+    Determine whether the nox session environment has been cached.
+
+    Parameters
+    ----------
+    session: object
+        A `nox.sessions.Session` object.
+    fname: str
+        Requirements filename that defines this environment.
+
+    Returns
+    -------
+    bool
+        Whether the session has been cached.
+
+    """
+    result = False
+    tmp_dir = Path(session.create_tmp())
+    python_version = session.python.replace(".", "")
+    cache = tmp_dir / f"py{python_version}.sha"
+    if cache.is_file():
+        with open(fname, "rb") as fi:
+            expected = hashlib.sha256(fi.read()).hexdigest()
```

Comment:
I'm not sure if the naming of `expected` vs `actual` is helpful to understanding here since what is 'expected' and what is 'actual' is a matter of perspective. You could also argue that the cache describes what you are expecting and the hash of the requirements is what you actually get.

Perhaps something more explicit like `hashed_requirements` and `cached_value` would be better.

Reply:
Sure